src/python/pants/backend/go/target_types.py

```diff
@@ -0,0 +1,46 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from pants.engine.rules import collect_rules
+from pants.engine.target import COMMON_TARGET_FIELDS, Dependencies, Sources, StringField, Target
+
+
+class GoSources(Sources):
+    expected_file_extensions = (".go",)
+
+
+class GoPackageSources(GoSources):
+    default = ("*.go", "!*_test.go")
+
+
+class GoImportPath(StringField):
+    # TODO: Infer the import path from the closest ancestor `go_module` target once that target is supported.
+    alias = "import_path"
+    help = """Import path in Go code to import this package."""
+
+
+class GoPackage(Target):
+    alias = "go_package"
+    core_fields = (*COMMON_TARGET_FIELDS, Dependencies, GoPackageSources, GoImportPath)
+    help = """A single Go package."""
+
+
+class GoBinaryName(StringField):
+    alias = "binary_name"
+    required = True
+    help = """Name of the Go binary to output."""
```

Comment:
This should probably be `OutputPathField`, which is the standardized way for changing the asset name for a target hooked up to `./pants package`. https://www.pantsbuild.org/docs/plugins-package-goal#1-set-up-a-package-target-type-recommended

Reply:
Fine to do that as a follow-up with a TODO here?

Reply:
Actually the field set already has an `output_path` field. That `output_path` is combined with `binary_name` to produce the final path.

Is `output_path` supposed to be the path to the file or to the directory in which the output is placed?

Reply:
> Actually the field set already has an output_path field

Note that you don't register it on the target, though, so the user has no way of setting the field. The FieldSet will look for that field, and fail to find it, so fall back to a default implementation.

> Is output_path supposed to be the path to the file or to the directory in which the output is placed?

Neither, it's the final path, which defaults to `path.to.parent/main.pex` for example.

Reply:
OutputPathField wants to append a file_ending, which even if blank, still appends a "." which would be incorrect for Go binaries:

```
        return self.value or os.path.join(
            address.spec_path.replace(os.sep, "."), f"{address.target_name}.{file_ending}"
        )
```

Reply:
This should be fixed in a follow-up PR.

Reply:
Okay, sounds good. Agreed that we should allow for no file ending.

Reply:
Added a TODO.